### PR TITLE
perf(gemm): split-K the router projection GEMV for decode occupancy (+3% decode)

### PR DESCRIPTION
**Pull request by Dexterity104**
## Summary

The decode MoE router projection (`launch_gemv_f32`, computing the per-expert logits from the post-attention hidden) ran on the one-warp-per-row `gemv_kernel`. With N = n_experts = 128 that launches only 128 warps, so the kernel sits far below the bandwidth roofline and a node-trace shows it at roughly 5.7% of the per-token decode. Two changes lift it: split-K so several warps cooperate per output row, and reading the small activation straight from L2 instead of staging it in shared with a `__syncthreads` (that barrier is a fixed per-block cost that dominates at this size).

## What changed

`gemv_f32_sk_kernel`: S warps per output row, each accumulating a 1/S stride of the K reduction in fp32, then an S-way reduce in shared. The activation is read as 128-bit `uint4` loads directly from global (L2-resident for the tiny router input), so there is no shared staging and no staging barrier. `launch_gemv_f32` routes through it with S=4 (RPB=2 rows per block) for the K-multiple-of-8 case; `SPARKINFER_ROUTER_SK=0` restores the plain one-warp-per-row kernel. Only the fp reduction order changes, so the result is numerically faithful.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Decode tok/s** (end-to-end, from `bench/scripts/bench.sh`, n=128, bs=1):

| | decode tok/s |
|---|--:|
| before (main) | 331.8 |
| after (this PR) | 341.9 |

About +3.0% decode, stable across three alternating runs. Accuracy gate vs llama.cpp (`bench/scripts/accuracy.sh`): top-1 `0.970`, KL `0.143` nats, matching the baseline (`0.970` / `0.151`) since the change only reorders the fp reduction. `ctest` 5/5, `compute-sanitizer --tool memcheck` clean (0 errors).

```text
# before (main path: SPARKINFER_ROUTER_SK=0)
decode tg    : 331.86 tok/s   (n=128, bs=1)
decode tg    : 331.86 tok/s   (n=128, bs=1)
decode tg    : 331.63 tok/s   (n=128, bs=1)
# after (this PR, default)
decode tg    : 341.85 tok/s   (n=128, bs=1)
decode tg    : 341.98 tok/s   (n=128, bs=1)
decode tg    : 341.86 tok/s   (n=128, bs=1)
```

### CODE REVIEW

**Comment by ai-hpc** [comment]
## ✅ sparkinfer auto-eval — `fc1b14f`

| metric | value |
|---|---|
| **label** | `eval:XS` |
| decode | 331.89 tok/s |
| vs prior frontier | 320.67 tok/s → +3.5% (+11.2) |
| **→ new frontier** | **331.89 tok/s** |
| correctness | top-1 97.0% · KL 0.1428 |

Verified speedup — **sets the new frontier to 331.89 tok/s** (was 320.67).

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by ai-hpc** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by ai-hpc** [comment]
Rebased onto current `main` (now includes #74 + the GPU-observability/thermal runtime). Your commits replayed cleanly with no conflicts — force-pushed on your behalf (maintainer edit). It'll be **re-evaluated for marginal gain** over the updated frontier on the next round.

**Comment by ai-hpc** [comment]
⚠️ **Correction:** the `eval:XL` posted moments ago was a **measurement artifact**, not a real result. Instance 42613343 was resumed *cold*, so its GPU clocks were still ramping when the same-box `main` baseline was measured (~285 tok/s) but at full boost when this PR was measured (~367 tok/s) — a ~29% clock-drift that even pushed the result *above the hardware ceiling* (impossible for this change). The bogus label + dashboard entry are reverted; this PR stays `needs-rebase` and will be re-evaluated on a thermally stable box.

**Comment by ai-hpc** [comment]
## ✅ sparkinfer auto-eval — `09bf7f9`

| metric | value |
|---|---|
| **label** | `eval:XS` |
| decode | 367.61 tok/s |
| vs prior frontier | 355.57 tok/s → +3.4% (+12.0) |
| **→ new frontier** | **367.61 tok/s** |
| correctness | top-1 99.0% · KL 0.1417 |

Verified speedup — **sets the new frontier to 367.61 tok/s** (was 355.57).

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by ai-hpc** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by ai-hpc** [comment]
## ✅ sparkinfer auto-eval — `9b4a876`

| metric | value |
|---|---|
| **label** | `eval:S` |
| decode | 384.08 tok/s |
| vs prior frontier | 370.9 tok/s → +3.6% (+13.2) |
| **→ new frontier** | **384.08 tok/s** |
| correctness | top-1 97.0% · KL 0.1417 |

Verified speedup — **sets the new frontier to 384.08 tok/s** (was 370.9).

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by ai-hpc** [comment]
The round's `merge-first` PR was just merged. Please **rebase this branch onto `main`** — once you push the rebase the bot re-evaluates it against the new frontier (crediting your *marginal* gain on top of what merged).

**Comment by Dexterity104** [comment]
Hi @ai-hpc Could you please review this PR again? Thanks!

**Comment by ai-hpc** [comment]
## ✅ sparkinfer auto-eval — `d459d58`

| metric | value |
|---|---|
| **label** | `eval:S` |
| decode | 394.45 tok/s |
| vs prior frontier | 381.04 tok/s → +3.5% (+13.4) |
| **→ new frontier** | **394.45 tok/s** |
| correctness | top-1 97.0% · KL 0.1417 |

Verified speedup — **sets the new frontier to 394.45 tok/s** (was 381.04).

_RTX 5090 (sm_120) · built from source · correctness vs llama.cpp. Automated — **not merged**; merge manually after review._

**Comment by ai-hpc** [comment]
Thanks for nice contribution @Dexterity104 
Can you please join discord to communicate better ?
https://discord.gg/8zYQRP92uQ
